File: minegame-game/src/main/java/top/nipuru/minegame/game/util/font/Font.java
package top.nipuru.minegame.game.util.font;

public class Font {
    public static final String DEFAULT = "default";

    private final String font;
    private final char character;
    private final float width, boldWidth;

    protected Font(String font, char character, float width, float boldWidth) {
        this.font = font;
        this.character = character;
        this.width = width;
        this.boldWidth = boldWidth;
    }

    public String getFont() {
        return font;
    }

    public char getCharacter() {
        return character;
    }

    public float getWidth() {
        return width;
    }

    public float getBoldWidth() {
        return boldWidth;
    }

    public static Font steve(char character, int width) {
        return new Font("default", character, width, width + 1);
    }

    public static Font steve(String font,char character, int width) {
        return new Font(font, character, width, width + 1);
    }

    public static Font slim(char character, float width) {
        return new Font(DEFAULT, character, width, width + 0.5f);
    }

    public static Font slim(String font,char character, float width) {
        return new Font(font, character, width, width + 0.5f);
    }

    public static Font font(char character, float width, float boldWidth) {
        return new Font(DEFAULT, character, width, boldWidth);
    }

    public static Font font(Bitmap bitmap, float width, float boldWidth) {
        return new Font(bitmap.getFont(), bitmap.getChar(), width, boldWidth);
    }

    public static Font[] steves(String string, int width) {
        Font[] fonts = new Font[string.length()];
        for (int i = 0; i < string.length(); i++) {
            fonts[i] = steve(string.charAt(i), width);
        }
        return fonts;
    }

    public static Font[] steves(Bitmap bitmap, int width) {
        Font[] fonts = new Font[bitmap.getSize()];
        for (int i = 0; i < bitmap.getSize(); i++) {
            fonts[i] = steve(bitmap.getChar(i), width);
        }
        return fonts;
    }

    public static Font[] slims(String string, float width) {
        Font[] fonts = new Font[string.length()];
        for (int i = 0; i < string.length(); i++) {
            fonts[i] = slim(string.charAt(i), width);
        }
        return fonts;
    }

    public static Font[] slims(Bitmap bitmap, float width) {
        Font[] fonts = new Font[bitmap.getSize()];
        for (int i = 0; i < bitmap.getSize(); i++) {
            fonts[i] = slim(bitmap.getChar(i), width);
        }
        return fonts;
    }

    public static Font[] fonts(String string, float width, float boldWidth) {
        Font[] fonts = new Font[string.length()];
        for (int i = 0; i < string.length(); i++) {
            fonts[i] = font(string.charAt(i), width, boldWidth);
        }
        return fonts;
    }

    public static Font[] fonts(Bitmap bitmap, float width, float boldWidth) {
        Font[] fonts = new Font[bitmap.getSize()];
        for (int i = 0; i < bitmap.getSize(); i++) {
            fonts[i] = new Font(bitmap.getFont(), bitmap.getChar(i), width, boldWidth);
        }
        return fonts;
    }
}

File: server/server-bukkit/src/main/kotlin/server/bukkit/util/text/Font.kt
package server.bukkit.util.text

import net.kyori.adventure.key.Keyed
import org.bukkit.NamespacedKey

class Font private constructor(val font: Keyed, val character: Char, val width: Float, val boldWidth: Float) {

    companion object {

        val DEFAULT: Keyed = NamespacedKey.minecraft("default")

        fun steve(font: Keyed = DEFAULT, character: Char, width: Int): Font {
            return Font(font, character, width.toFloat(), (width + 1).toFloat())
        }

        fun steves(font: Keyed = DEFAULT, string: String, width: Int): List<Font> {
            val fonts = ArrayList<Font>(string.length)
            for (element in string) { fonts += steve(font, element, width) }
            return fonts
        }

        fun steves(bitmap: Bitmap, width: Int): List<Font> {
            return steves(bitmap.font, bitmap.string, width)
        }

        fun slim(font: Keyed = DEFAULT, character: Char, width: Float): Font {
            return Font(font, character, width, width + 0.5f)
        }

        fun slims(font: Keyed = DEFAULT, string: String, width: Float): List<Font> {
            val fonts = ArrayList<Font>(string.length)
            for (element in string) { fonts += slim(font, element, width) }
            return fonts
        }

        fun slims(bitmap: Bitmap, width: Float): List<Font> {
            return slims(bitmap.font, bitmap.string, width)
        }

        fun font(font: Keyed = DEFAULT, character: Char, width: Float, boldWidth: Float): Font {
            return Font(font, character, width, boldWidth)
        }

        fun fonts(font: Keyed = DEFAULT, string: String, width: Float, boldWidth: Float): List<Font> {
            val fonts = ArrayList<Font>(string.length)
            for (element in string) { fonts += font(font, element, width, boldWidth) }
            return fonts
        }

        fun fonts(bitmap: Bitmap, width: Float, boldWidth: Float): List<Font> {
            return fonts(bitmap.font, bitmap.string, width, boldWidth)
        }
    }
}
